### PR TITLE
Public API for constructing NT with jagged layout from tensor list

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111078
* #109123

This PR:

* Introduces a new layout: torch.jagged
* Expands torch.nested.nested_tensor(tensor_list) to accept a layout kwarg with the default set to torch.strided

TODO: non-copying constructor

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

File: torch/nested/__init__.py [hunk 1]
```diff
@@ -116,11 +116,8 @@ def as_nested_tensor(
 """,
 )
 
-nested_tensor = _add_docstr(
-    _nested.nested_tensor,
+def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires_grad=False, pin_memory=False) -> Tensor:
```
Comment (cpuhrsch) [inline torch/nested/__init__.py:119]: You'll need to update the `layout=None` or catch it below.

File: torch/nested/__init__.py [hunk 2]
```diff
@@ -145,5 +144,21 @@ def as_nested_tensor(
...
+        return _nested.nested_tensor(
+            tensor_list,
+            dtype=dtype,
+            device=device,
+            requires_grad=requires_grad,
+            pin_memory=pin_memory)
+    elif layout == torch.jagged:
+        from torch.nested._internal.nested_tensor import jagged_from_list
+
+        nt = jagged_from_list(tensor_list).requires_grad_(requires_grad)
```
Comment (cpuhrsch) [inline torch/nested/__init__.py:158]: I think jagged_from_list needs to have offsets=None.

**Review by soulitzer — APPROVED** [review]
SHIP

**Comment by jcaip** [comment]
Sorry @jbschlosser, I forgot I rebased on top of this PR when I force pushed.

**Pull request closed without merging**